framework/quota/test/org/apache/cloudstack/quota/QuotaStatementTest.java

```diff
@@ -0,0 +1,248 @@
+
+import javax.naming.ConfigurationException;
+
+import java.lang.reflect.Field;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.List;
+
+@RunWith(MockitoJUnitRunner.class)
+public class QuotaStatementTest extends TestCase {
+
+    @Mock
+    AccountDao accountDao;
+    @Mock
+    QuotaAccountDao quotaAcc;
+    @Mock
+    ConfigurationDao configDao;
+    @Mock
+    QuotaUsageDao quotaUsage;
+    @Mock
+    QuotaAlertManager emailQuotaAlert;
+
+    @Spy
+    QuotaStatementImpl quotaStatement = new QuotaStatementImpl();
+
+    private void injectMockToField(Object mock, String fieldName) throws NoSuchFieldException, IllegalAccessException {
+        Field f = QuotaStatementImpl.class.getDeclaredField(fieldName);
+        f.setAccessible(true);
+        f.set(quotaStatement, mock);
+    }
+
+    @Before
+    public void setup() throws IllegalAccessException, NoSuchFieldException, ConfigurationException {
+        // Dummy transaction stack setup
+        TransactionLegacy.open("QuotaStatementImplTest");
+
+        injectMockToField(accountDao, "_accountDao");
+        injectMockToField(quotaAcc, "_quotaAcc");
+        injectMockToField(configDao, "_configDao");
+        injectMockToField(quotaUsage, "_quotaUsage");
+        injectMockToField(emailQuotaAlert, "_quotaAlert");
+    }
+
+    @Test
+    public void testStartStop() {
+        try {
+            quotaStatement.start(); // expected to fail as pid is not available
+        } catch (NumberFormatException ignored) {
+        }
+        assertTrue(quotaStatement.stop());
+    }
+
+    @Test
+    public void testStatementPeriodBIMONTHLY() {
+        Calendar date = Calendar.getInstance();
+
+        //BIMONTHLY - first statement of month
+        date.set(Calendar.DATE, QuotaStatementImpl.s_LAST_STATEMENT_SENT_DAYS + 1);
+        Calendar period[] = quotaStatement.statementTime(date, STATEMENT_PERIODS.BIMONTHLY);
+        assertTrue(period == null);
+
+        //1 of this month
+        date.set(Calendar.DATE, 1);
+        period = quotaStatement.statementTime(date, STATEMENT_PERIODS.BIMONTHLY);
+        assertTrue(period != null);
+        assertTrue(period[0].toString(), period[0].before(period[1]));
+        assertTrue(period[0].toString(), period[0].get(Calendar.DATE) == 1);
+        assertTrue(period[1].toString(), period[1].get(Calendar.DATE) == 15);
+
+        //BIMONTHLY - second statement of month
+        date = Calendar.getInstance();
+        date.set(Calendar.DATE, QuotaStatementImpl.s_LAST_STATEMENT_SENT_DAYS + 16);
+        period = quotaStatement.statementTime(date, STATEMENT_PERIODS.BIMONTHLY);
+        assertTrue(period == null);
+
+        //17 of this month
+        date.set(Calendar.DATE, 17);
+        period = quotaStatement.statementTime(date, STATEMENT_PERIODS.BIMONTHLY);
+        assertTrue(period != null);
+        assertTrue(period[0].toString(), period[0].before(period[1]));
```

Comment:
Add an assertion that the length of ``period`` is 2 to avoid a confusing ``ArrayIndexOutofBoundsException``.